File: src/main/java/com/hileco/cortex/primitives/layer/LayeredVmStack.java
package com.hileco.cortex.primitives.layer;

import com.hileco.vm.primitives.VmStack;

public class LayeredVmStack<T> implements VmStack<T, LayeredVmStack<T>> {
    private StackLayer<T> edge;

    public LayeredVmStack() {
        this(new StackLayer<>(null));
    }

    private LayeredVmStack(StackLayer<T> edge) {
        this.edge = edge;
    }

    public T peek() {
        return get(edge, edge.getLength() - 1);
    }

    private T get(StackLayer<T> layer, int index) {
        if (index >= layer.getLength()) {
            return null;
        }
        var layerOffset = layer.getLength() - layer.getEntries().size();
        var layerIndex = index - layerOffset;
        if (layerIndex >= 0 && layerIndex < layer.getEntries().size()) {
            return layer.getEntries().get(layerIndex);
        }
        var parent = layer.getParent();
        if (parent == null) {
            return null;
        }
        return get(parent, index);
    }

    public void push(T value) {
        edge.getEntries().push(value);
        edge.setLength(edge.getLength() + 1);
    }

    public T pop() {
        var value = peek();
        if (!edge.getEntries().isEmpty()) {
            edge.getEntries().pop();
        }
        var length = edge.getLength();
        if (length > 0) {
            edge.setLength(length - 1);
        }
        return value;
    }

    public int size() {
        return 0;
    }

    public LayeredVmStack<T> copy() {
        var child1 = new StackLayer<>(edge);
        var child2 = new StackLayer<>(edge);
        this.edge = child1;
        return new LayeredVmStack<>(child2);
    }

    public void close() {
        edge.close();
    }
}

File: cortex-data/src/main/kotlin/com/hileco/cortex/data/layer/LayeredVmStack.kt
package com.hileco.cortex.data.layer


import com.hileco.cortex.data.base.BaseVmStack

class LayeredVmStack<V> : BaseVmStack<V, LayeredVmStack<V>> {
    private var edge: StackLayer<V>

    private constructor(edge: StackLayer<V>) {
        this.edge = edge
    }

    constructor() {
        this.edge = StackLayer(null)
    }

    override fun push(value: V) {
        this.edge.entries[this.edge.length] = value
        this.edge.length++
    }

    override fun pop(): V {
        val value = peek()
        this.edge.entries.remove(this.edge.length - 1)
        this.edge.length--
        return value
    }

    override fun peek(offset: Int): V {
        return this[this.edge.length - offset - 1]
    }

    override operator fun get(index: Int): V {
        var chosen: StackLayer<V>? = this.edge
        while (chosen != null && chosen.length > index) {
            val value = chosen.entries[index]
            if (value != null) {
                return value
            }
            chosen = chosen.parent
        }
        throw IndexOutOfBoundsException("size ${this.edge.length} <= index $index")
    }

    override operator fun set(index: Int, value: V) {
        if (this.edge.length < index) {
            throw IndexOutOfBoundsException("size ${this.edge.length} <= index $index")
        }
        this.edge.entries[index] = value
    }

    override fun size(): Int {
        return this.edge.length
    }

    override fun clear() {
        this.edge.entries.clear()
        this.edge.length = 0
    }

    override fun close() {
        this.edge.close()
    }

    override fun copy(): LayeredVmStack<V> {
        while (edge.isEmpty) {
            this.edge = edge.parent ?: return LayeredVmStack()
        }
        val child1 = StackLayer(edge)
        val child2 = StackLayer(edge)
        this.edge = child1
        return LayeredVmStack(child2)
    }
}
